### DISCUSSION TITLE
Repeated access check when using custom GraphQL directive for column level masking

### DISCUSSION THREAD

**Discussion by Kiki-ooops** [graphql-java/graphql-java#3114] [General]
Hi, I'm trying to implement a custom GraphQL directive for column level access control. Below is my GraphQL schema with directive definition:

```
directive @auth(role : String!) on FIELD_DEFINITION
type Item{
	id: Int @auth(role : "admin")
}
```

This directive is used to add access control on field level. Below is my custom directive implementation, which uses `columnLevelSecurityService.hasRole()` to validate if the user has access to the requested field.

```
public class AuthorizationDirective implements SchemaDirectiveWiring {
    .....

    @Override
    public GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> environment) {
        GraphQLDirective directive = environment.getDirective();
        if (directive != null) {
            String targetAuthRole = environment.getDirective().getArgument("role").toString();

            GraphQLFieldDefinition field = environment.getElement();
            GraphQLFieldsContainer parentType = environment.getFieldsContainer();

            DataFetcher originalDataFetcher = environment.getCodeRegistry().getDataFetcher(parentType, field);
            DataFetcher authDataFetcher = dataFetchingEnvironment -> {
                if (columnLevelSecurityService.hasRole(targetAuthRole)) {
                    return originalDataFetcher.get(dataFetchingEnvironment);
                } else {
                    return null;
                }
            };

            environment.getCodeRegistry().dataFetcher(parentType, field, authDataFetcher);
            return field;
        }
        .....
    }
}
```
However, when I run my query, I noticed the `hasRole` method has been called for every single data row returned by the query. Calling `hasRole()` could be expensive due to our implementation, and we'd like to avoid calling it too frequently. Is it possible to call `hasRole` method and do access check only once for each query request?

Thanks!

**Comment by bbakerman** [comment]
You have said its call N times but you have not listed your schema that leads to and `Item`

I am guessing its something like 

```graphql
type Query {
   getItems(args : X) : [Item!]
}
```

and hence the query is

```
query q 
  getItems(args : ...) {
    id
    desc
  }
  }
```

What happens here is that the `getItems` data fetcher returns a list if items and then for each item in the list a data fetcher is called for each property - in your case `id` - so if there are 10 items found - a PropertyDatafetcher is called 10 times for each id.

I think you need to move the logic up a level into the `getItems` field. That is fetch a list of things and before giving them back - inspect each object and check the user has permission.  

Am I far off on the `getItems` guess?  If you expand your schema I might be able to help more

**Comment by Kiki-ooops** [comment]
> 

Hi, below is my simplified schema based on your suggestion:
```
directive @enableColumnLevelMasking on FIELD_DEFINITION
directive @auth(role : String!) on FIELD_DEFINITION

type Query{
      getAllItems(): [Item] @enableColumnLevelMasking
}

type Item{
	id: Int
	name: String @auth(role : "admin")
        ....
}
```
And the simplified query goes like:
```
{
  getAllItems() {
    id
    name
  }
}
```
My understanding is: in the lambda expression of the `dataFetchingEnvironment`, first grab all the fields that the user want to access and store them in a list. Then grab all fields that need the access check and store them in a map. The key of this map is the field name, the value is the required role. If the field that the user want to access exists in the keys of the map, I would call the `hasRole()` method. Below is my "pseudocode" for this:

```
public GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> environment) {
     ......
     DataFetcher authDataFetcher = dataFetchingEnvironment -> {

          // grab the fields that the user want to access
          List<SelectedField> selectedFields = dataFetchingEnvironment.getSelectionSet().getImmediateFields();

         // declare a map to store the fields that has access control, key is the field name, value is the required role
          Map<String , String> map = new HashMap<>();

          Grab all the fields that the item has and store them in a list L

          For field in L
               if field has access control
                    put field name and required role to the map
                
          For selectedField in selectedFields
               if map containsKey selectedField
                    call hasRole()
          .....
     };
     .....
}
```
Is this correct? In this way, it should only check once for the column I add access control on. 

Appreciate your help!
